ddl/partition.go

```diff
@@ -2388,35 +2391,58 @@ func (w *worker) onExchangeTablePartition(d *ddlCtx, t *meta.Meta, job *model.Jo
 		return ver, errors.Trace(err)
 	}
 
-	if pt.State != model.StatePublic {
-		job.State = model.JobStateCancelled
-		return ver, dbterror.ErrInvalidDDLState.GenWithStack("table %s is not in public, but %s", pt.Name, pt.State)
-	}
-
-	err = checkExchangePartition(pt, nt)
+	index, partDef, err := getPartitionDef(pt, partName)
 	if err != nil {
-		job.State = model.JobStateCancelled
 		return ver, errors.Trace(err)
 	}
+	if job.SchemaState == model.StateNone {
```

Comment:
If there is a running Exchange Partition from previous version, there is no harm in redoing this step and set job.SchemaState and reset ExchangePartitionInfo.